api/services/update/steps.js

```diff
@@ -0,0 +1,309 @@
+  let readStream;
+  try {
+    readStream = fs.createReadStream(filePath);
+  } catch (err) {
+    logger.error(`fs.createReadStream in insertDataUnpaywall: ${err}`);
+    await fail(stateName);
+    // TODO throw Error
+  }
+
+  // get information "loaded" for state
+  let loaded = 0;
+  readStream.on('data', (chunk) => {
+    loaded += chunk.length;
+  });
+
+  let decompressedStream;
+  try {
+    decompressedStream = readStream.pipe(zlib.createGunzip());
+  } catch (err) {
+    logger.error(`readStream.pipe(zlib.createGunzip()) in insertDataUnpaywall: ${err}`);
+    await fail(stateName);
+    // TODO throw Error
+  }
+
+  const rl = readline.createInterface({
+    input: decompressedStream,
+    crlfDelay: Infinity,
+  });
+
+  // array that will contain the packet of 1000 unpaywall data
+  let tab = [];
+
+  // Reads line by line the output of the decompression stream to make packets of 1000
+  // to insert them in bulk in an elastic
+  for await (const line of rl) {
+    // limit
+    if (step.linesRead === opts.limit) {
+      break;
+    }
+
+    step.linesRead += 1;
+
+    // offset
+    if (step.linesRead >= opts.offset + 1) {
+      // fill the array
+      try {
+        tab.push(JSON.parse(line));
+      } catch (err) {
+        logger.error(`JSON.parse in insertDataUnpaywall: ${err}`);
+        await fail(stateName);
+        // TODO throw Error
+      }
+    }
+    // bulk insertion
+    if (tab.length % 1000 === 0 && tab.length !== 0) {
+      await insertUPW(tab, index);
+      step.percent = ((loaded / bytes.size) * 100).toFixed(2);
+      step.took = (new Date() - start) / 1000;
+      state.steps[state.steps.length - 1] = step;
+      await updateStateInFile(state, stateName);
+      tab = [];
+    }
+    if (step.linesRead % 100000 === 0) {
+      logger.info(`${step.linesRead} Lines reads`);
+    }
+  }
+  // last insertion if there is data left
+  if (tab.length !== 0) {
+    await insertUPW(tab, index);
+    tab = [];
+  }
+  logger.info('step - end insertion');
+
+  // last update of step
+  step.status = 'success';
+  step.took = (new Date() - start) / 1000;
+  step.percent = 100;
+  state.steps[state.steps.length - 1] = step;
+  await updateStateInFile(state, stateName);
+};
+
+/**
+ * Update the step the percentage in download regularly until the download is complete
+ * @param {setTimeout} timeout - timeout
+ * @param {string} filePath - path where the file is downloaded
+ * @param {object} info - info of file
+ * @param {string} stateName - state filename
+ * @param {object} state - state in JSON format
+ * @param {date} start - download start date
+ */
+async function updatePercentStepDownload(timeout, filePath, info, stateName, state, start) {
+  const step = state.steps[state.steps.length - 1];
+  let bytes;
+  try {
+    bytes = await fs.stat(filePath);
+  } catch (err) {
+    logger.error(`fs.stat in updatePercentStepDownload: ${err}`);
+  }
+  if (bytes.size === info.size) {
+    clearTimeout(timeout);
```

Comment:
There's no need to cancel the timeout outside of the actual timeout function, so this is not required.